### PR TITLE
Enable GetWorkflowExecutionHistory after mutable state is deleted

### PR DESCRIPTION
**Pull request by tamer-eldeeb**
In this change we keep the length of the history in the visibility record.
This is useful in itself, but also allows reading the execution history even after deleting its mutable state.
Issue #206

### CODE REVIEW

File: service/frontend/handler.go [hunk 1]
```diff
@@ -558,11 +558,7 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
 		return nil, errWorkflowIDNotSet
 	}
 
-	if !getRequest.GetExecution().IsSetRunId() {
-		return nil, errRunIDNotSet
-	}
-
-	if uuid.Parse(getRequest.GetExecution().GetRunId()) == nil {
+	if getRequest.GetExecution().IsSetRunId() && uuid.Parse(getRequest.GetExecution().GetRunId()) == nil {
```
Comment (venkat1109) [inline service/frontend/handler.go]: this code makes me think you are making runID param optional. Is that the intention ? If not, you still need the if !setRunID { return err} check

File: schema/visibility/schema.cql [hunk 2]
```diff
@@ -25,6 +25,7 @@ CREATE TABLE closed_executions (
   close_time           timestamp,
   status               int,  -- enum WorkflowExecutionCloseStatus {COMPLETED, FAILED, CANCELED, TERMINATED, CONTINUED_AS_NEW, TIMED_OUT}
   workflow_type_name   text,
+  history_length       bigint,
```
Comment (venkat1109) [inline schema/visibility/schema.cql]: also add the new field inside schema/visibility/versioned/...

File: service/frontend/handler.go [hunk 3]
```diff
@@ -587,11 +583,25 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
...
+		if err == nil {
+			token.nextEventID = response.GetEventId()
+			token.runID = response.GetRunId()
+		} else {
+			if _, ok := err.(*gen.EntityNotExistsError); !ok || !getRequest.GetExecution().IsSetRunId() {
+				return nil, wrapError(err)
+			}
+			// It is possible that we still have the events in the table even though the mutable state is gone
+			// Get the nextEventID from visibility store if we still have it.
+			visibilityResp, err := wh.visibitiltyMgr.GetClosedWorkflowExecution(&persistence.GetClosedWorkflowExecutionRequest{
```
Comment (venkat1109) [inline service/frontend/handler.go]: I believe this call is used by both decision workers as well as by the visibility client (CLI). Do you want to make the assumption here that the caller is always interested in the ClosedWFExecution() ? For decision workers, this API call must still return an error if the WFExecution does not exist. How about passing in a special flag in the request struct, that gives you a hint that the caller is interested in closedWFExecutions as well ?

Comment (tamer-eldeeb) [inline service/frontend/handler.go]: This path will never be hit by decision workers. If there is a decision, the workflow is not deleted. Also, they get a token from the PollForDecisionTask call anyway so we know what the nextEventId is from the token.

File: service/frontend/handler.go (hunk 1 above)
Comment (tamer-eldeeb) [inline service/frontend/handler.go]: It is intentional, to allow getting history of running execution using the workflow_id only

File: common/persistence/cassandraVisibilityPersistence.go [hunk 4]
```diff
@@ -446,6 +454,35 @@ func (v *cassandraVisibilityPersistence) ListClosedWorkflowExecutionsByStatus(
...
+		execution.GetRunId())
+
+	iter := query.Iter()
+	if iter == nil {
+		return nil, &workflow.InternalServiceError{
+			Message: "GetClosedWorkflowExecution operation failed.  Not able to create query iterator.",
+		}
+	}
+
+	wfexecution, has := readClosedWorkflowExecutionRecord(iter)
```
Comment (samarabbas) [inline common/persistence/cassandraVisibilityPersistence.go]: Make sure to call iter.Close(), this should return back an error which we could wrap into InternalServerError and return back to application.

**Pull request merged**